### DISCUSSION TITLE
Controlling devices via a shift register

### DISCUSSION THREAD

**Discussion by Seggan** [mattjlewis/diozero#77] [Q&A] [answered]
Is there any way to use the built in classes (i.e. `PwmLed`) via a shift register? I don't want to reinvent the wheel just so I can use a shift register chip with my code. Here's my shift register code btw:
```java
/**
 * A shift register chip designed for output, i.e. the 74HC595
 */
public class OutputShiftRegister implements DeviceInterface {

    private final DigitalOutputDevice data;
    private final DigitalOutputDevice dataWrite;
    private final DigitalOutputDevice dataOut;

    private boolean[] buf;

    private boolean flushOnWrite;

    public OutputShiftRegister(int width, int ser, int srclk, int rclk, boolean flushOnWrite) {
        this.data = new DigitalOutputDevice(ser);
        this.dataWrite = new DigitalOutputDevice(srclk);
        this.dataOut = new DigitalOutputDevice(rclk);

        this.buf = new boolean[width];
        this.flushOnWrite = flushOnWrite;
    }

    public OutputShiftRegister(int width, int ser, int srclk, int rclk) {
        this(width, ser, srclk, rclk, true);
    }

    /**
     * Writes the value to the internal buffer. If {@link #flushOnWrite()} is
     * {@code true}, the the buffer will then be written to the chip
     *
     * @throws IndexOutOfBoundsException if the index is bigger than what the chip can hold
     */
    public void write(int index, boolean value) {
        try {
            buf[index] = value;
        } catch (ArrayIndexOutOfBoundsException e) {
            throw new IndexOutOfBoundsException(index);
        }

        if (flushOnWrite) {
            flush();
        }
    }

    /**
     * Writes the value directly to the chip instead of the buffer
     */
    public void writeValue(boolean value) {
        data.setOn(value);
        dataWrite.on();
        SleepUtil.sleepNanos(0, 1000);
        dataWrite.off();
    }

    /**
     * Writes the contents of the buffer to the chip and outputs it
     */
    public void flush() {
        for (boolean b : buf) {
            writeValue(b);
        }

        writeRegister();
    }

    /**
     * Clears the chip and the internal buffer
     */
    public void clear() {
        buf = new boolean[buf.length];

        for (int i = 0; i < buf.length; i++) {
            writeValue(false);
        }

        writeRegister();
    }

    /**
     * Writes the contents of the register in the chip to the chip's output
     */
    public void writeRegister() {
        dataOut.on();
        SleepUtil.sleepNanos(0, 1000);
        dataOut.off();
    }

    public boolean flushOnWrite() {
        return flushOnWrite;
    }

    public void setFlushOnWrite(boolean flushOnWrite) {
        this.flushOnWrite = flushOnWrite;
    }

    @Override
    public void close() throws RuntimeIOException {
        clear();
        this.data.close();
        this.dataWrite.close();
        this.dataOut.close();
    }
}

```

**Comment by mattjlewis** [comment, marked as answer]
Apologies, I've not done anything with shift registers before. Can you explain a bit more about how this might be implemented please.

I've just read this article https://www.arduino.cc/en/Tutorial/Foundations/ShiftOut and can sort of understand the GPIO digital output multiplexing. The way to handle that would be to create a shift register digital output device factory similar to the MCP23007 class. Let me have a think but I'd be interested in your thoughts.

**Reply by Seggan** [reply]
So, the 74HC595 (the kind I'm using) has 3 input pins: a data pin, a clock pin, and a write pin. There are also a clear and a disable pin but those are not important for this discussion. There are 8 output pins on this specific chip. Suppose I have a brand-new, unmodified, cleared chip with 8 outputs. I set the data pin to the value I want, then set the clock pin to high. This writes the value at the data pin to slot 8 in the storage register. Set the clock pin to low, and then repeat for slots, 7, 6, 5, 4, 3, 2, and 1. Then set the write pin to high, and  the bytes that you wrote are outputted to 8 output pins on the chip. I saw a cool animation of this somewhere but I can't find it... 

If you need I'll be happy to help write the code.

\- Seggan

**Reply by Seggan** [reply]
> I saw a cool animation of this somewhere but I can't find it...

Got it: https://lastminuteengineers.com/wp-content/uploads/arduino/74HC595-Shift-Register-Working.gif

**Comment by Seggan** [comment]
As a side note, the `buf` variable I have in my code is simply to edit the bits without writing to the chip and then once I wrote them flush em all in one write. It has no relation to the operation of the chip.

**Reply by mattjlewis** [reply]
I think I get I now, I’ve ordered a couple of the 74HC595 ones. The [GpioDeviceFactoryInterface](https://github.com/mattjlewis/diozero/blob/main/diozero-core/src/main/java/com/diozero/internal/spi/GpioDeviceFactoryInterface.java) should help here. As mentioned, the MCP23007 is a good example to see how this works - actual implementation is here: https://github.com/mattjlewis/diozero/blob/main/diozero-core/src/main/java/com/diozero/devices/mcp23xxx/MCP23xxx.java.
Obviously this would only need to support digital output.
If I get a moment, I’ll take a look tomorrow.

**Reply by Seggan** [reply]
Great 👍 I'll see if I can work up a skeletal implementation.

**Reply by Seggan** [reply]
Is there another way we can converse? Like Discord or something? GH discussions isn't really great for certain kinds of stuff.

**Reply by mattjlewis** [reply]
I had a go at it, see [OutputShiftRegisterDeviceFactory](https://github.com/mattjlewis/diozero/blob/main/diozero-core/src/main/java/com/diozero/devices/sandpit/OutputShiftRegisterDeviceFactory.java) and associated [test app](https://github.com/mattjlewis/diozero/blob/main/diozero-sampleapps/src/main/java/com/diozero/sampleapps/sandpit/ShiftRegisterTest.java). I've tested it on an Arduino Uno using my Firmata provider (as it's convenient for developing directly on my laptop).
Need to try it with PwmLed now, not sure what the performance will be like.

**Reply by mattjlewis** [reply]
BTW setup a Discord channel, believe I sent you an invite...

**Reply by Seggan** [reply]
Theres PMs in Github? I didn't get anything

**Reply by Seggan** [reply]
If anything my discord is `Seggan#8111`
